### PR TITLE
Generate JSON data for packages and store it as an artifact

### PR DESCRIPTION
**Pull request by maximedenes**
CI setup still to be tested.

### CODE REVIEW

**Comment by maximedenes** [comment]
Failures are orthogonal, and due to https://github.com/coq/opam-coq-archive/issues/799

File: scripts/archive2web.ml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  let versions = List.map (do_one_package_version root name) versions in
+  let pkg = { versions; most_recent } in
+  StrMap.update name (function None -> Some pkg | Some pkg' -> Some (merge_package_versions pkg' pkg)) pkgs
+
+let get_version pname pdir = strip_prefix (pname ^ ".") (Filename.basename pdir)
+
+let do_one_suite pkgs root =
+  fold_subdirs (root ^ "/packages/") pkgs (fun pkgs pname ->
+      let versions =
+        fold_subdirs (root ^ "/packages/" ^ pname ^ "/") [] (fun versions pdir ->
```
Comment (Zimmi48) [inline scripts/archive2web.ml]: Nitpick: why not use `/` here?

File: scripts/archive2web.ml [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            get_version pname pdir :: versions)
+      in
+      do_one_package pkgs ~versions root pname)
+
+let () =
+  let roots = parse_args Sys.argv in
+  let packages = List.fold_left do_one_suite StrMap.empty roots in
+  print_endline @@ Yojson.Safe.to_string @@ packages_to_yojson @@ StrMap.bindings packages
```
Comment (Zimmi48) [inline scripts/archive2web.ml]: Shamelessly nitpicking:
```ocaml
packages |> StrMap.bindings |> packages_to_yojson |> Yojson.Safe.to_string |> print_endline
```

reads much better IMHO.

File: scripts/dune [hunk 3]
```diff
@@ -0,0 +1,9 @@
+(install
+ (section bin)
+ (package opam-coq-archive)
+ (files (archive2web.exe as archive2web)))
```
Comment (Zimmi48) [inline scripts/dune]: I don't get the purpose of this `install` stanza. If you give a `public_name` and a `package` to the `executable` stanza below, it should be enough (see https://dune.readthedocs.io/en/latest/dune-files.html#executable).

File: scripts/opamVersionCompare.mli [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+(** {2 Comparing debian version strings} *)
+
+(** The following functions compare any two strings, that is these
+    functions do not check whether the arguments are really legal
+    debian versions. If the arguments are debian version strings, then
+    the result is as required by debian policy. Note that two strings
+    may be equivalent, that is denote the same debian version, even
+    when they differ in syntax, as for instance "0:1.2.00" and
+    "1.02-0".
+*)
```
Comment (Zimmi48) [inline scripts/opamVersionCompare.mli]: It seems a bit sad to have to duplicate this code even if we already depend on opam. I suppose opam does not expose any public ML API giving access to these functions? Maybe this should be requested to them.

File: scripts/archive2web.ml (hunk 2 above)
Comment (maximedenes) [reply scripts/archive2web.ml]: No way.

File: scripts/archive2web.ml (hunk 1 above)
Comment (maximedenes) [reply scripts/archive2web.ml]: Yep, this dates back to the original code from @gares. Will update.

File: scripts/opamVersionCompare.mli (hunk 4 above)
Comment (maximedenes) [reply scripts/opamVersionCompare.mli]: Does our build depend on opam?

File: scripts/archive2web.ml (hunk 2 above)
Comment (ppedrot) [reply scripts/archive2web.ml]: A new approach to polarization: left vs. right-leaning application?

File: scripts/dune (hunk 3 above)
Comment (maximedenes) [reply scripts/dune]: Fixed, thanks.

File: scripts/opamVersionCompare.mli (hunk 4 above)
Comment (Zimmi48) [reply scripts/opamVersionCompare.mli]: Oh maybe not.

File: default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+with pkgs;
+
+stdenv.mkDerivation rec {
+
+  name = "coq_website";
+  src = null;
+
+  buildInputs = [
+    dune
+    lua51Packages.lua
```
Comment (gares) [inline default.nix]: Is my beloved Lua still used?

File: Makefile [hunk 6]
```diff
@@ -23,16 +23,19 @@ run: all
 	$(H)cd www && python -m SimpleHTTPServer 8000
 
 check-deps: \
-	which-opam which-lua5.1 opam-config which-markdown yamlpp
+	which-opam which-markdown \
+	opkg-opam-file-format opkg-ocamlfind \
+	yamlpp scripts/archive2web.native
```
Comment (gares) [inline Makefile]: here you check for `archive2web.native` but the `all:` target up there calls just `archive2web`

File: scripts/archive2web.ml (hunk 2 above)
Comment (gares) [reply scripts/archive2web.ml]: why not
```
let json = packages |> StrMap.bindings |> packages_to_yojson in
print_endline @@ Yojson.Safe.to_string json
```

File: Makefile (hunk 6 above)
Comment (gares) [reply Makefile]: also, given that we use `dune exec` in CI I guess we should use it also in `all:`

Comment (gares) [reply Makefile]: finally, given you use nix to express requirements, you should probably remove this check in the first place

File: default.nix (hunk 5 above)
Comment (maximedenes) [reply default.nix]: Yes, because the `archive2web` script is still there for now.

**Comment by ejgallego** [comment]
Is this data specific to Coq's OPAM repository?

**Comment by maximedenes** [comment]
> Is this data specific to Coq's OPAM repository?

Sorry, I don't understand the question.

**Comment by ejgallego** [comment]
> Sorry, I don't understand the question.

I mean if the data format you are generating is exclusive to the Coq's OPAM repository, or it is meant to be used by the main OPAM repos.

**Comment by maximedenes** [comment]
> I mean if the data format you are generating is exclusive to the Coq's OPAM repository, or it is meant to be used by the main OPAM repos.

It is meant to be used by the Coq OPAM repository.

**Comment by ejgallego** [comment]
> It is meant to be used by the Coq OPAM repository.

Wouldn't it make sense to do this then in coordination with the main OPAM repos maintainers so the data export is something standard in the OPAM ecosystem?

I know the OPAM maintainers are developing their own advanced infrastructure so they may already have something similar or planning to do so.

If you are interested folks I can tell them to stop by this PR and comment.

**Comment by ejgallego** [comment]
For example http://check.ocamllabs.io/ would be pretty useful for the Coq repos too.

**Comment by maximedenes** [comment]
Feel free to talk to the OPAM developers, but for now, I'll pursue this way.

As for checking the build, I was planning to do it with Nix, as the OPAM constraint-based approach does not scale for that purpose. But it is another topic.

**Comment by maximedenes** [comment]
> Wouldn't it make sense to do this then in coordination with the main OPAM repos maintainers so the data export is something standard in the OPAM ecosystem?

Note that eventually, we may want to go the other way and generate the OPAM (and Nix) packages from the data.

**Comment by clarus** [comment]
For Coq I am maintaining https://coq-bench.github.io/ to check the install of the packages.

**Comment by ejgallego** [comment]
Hey @clarus , indeed that's very cool, for some reason that's not fully integrated with the Coq's OPAM repos, right? What do you think should be done here?

Do you also think it would make sense to converge with http://check.ocamllabs.io/ ? We can chat on Gitter if you want. The nice thing of check.ocamllabs is that it post to Slack a diff of buildability status, maybe something like that could be done?

**Comment by Zimmi48** [comment]
Indeed @clarus, that's a great tool, and I never really understood why it didn't have a more official status (in particular, it is hard to discover when you don't know about it). I agree with @ejgallego: it doesn't make sense that the OCaml community in the large sense is reinventing the same things over and over. I mentioned https://coq-bench.github.io/ in https://discuss.ocaml.org/t/check-ocamllabs-io-new-features-and-call-to-rename/3373 but did not get any reaction. Once again, I agree with @ejgallego that we should try to make sure to coordinate with the OCaml community and develop tools that can be useful to all. Obviously, this does not prevent experimenting on our side first, especially given that our objectives are not necessarily all aligned, but we shouldn't forget to contribute back.

**Comment by ejgallego** [comment]
> I mentioned https://coq-bench.github.io/ in https://discuss.ocaml.org/t/check-ocamllabs-io-new-features-and-call-to-rename/3373 but did not get any reaction.

Sure I can help getting @clarus and @kit-ty-kate in touch, but I'd be good to know first what @clarus has in mind for the tool.

Regarding the wider OPAM efforts I'm not sure given the response to my suggestion that is a good idea to involve them in this.

**Comment by clarus** [comment]
Thanks! Yes, this tool is maybe hard to discover. My aim is to have 100% of Coq packages able to install from a fresh repository with a given Coq version. When I have time, I take some errors and fix the meta-data of the failing package. Currently hoping to have some new errors fixed by using OCaml from the `opam switch` rather than from the system.

I started this project at the end of 2014 and have maintained it since then running it on my server. At that time I do not think I have heard about alternatives on the OCaml side.

Most of the development time was invested into making the platform reliable. I like it like that but it would probably need a notification system, like in Slack as you said @ejgallego .

**Comment by clarus** [comment]
We could probably use http://check.ocamllabs.io/ for Coq with some adaptations, but I like the stability and independence of https://coq-bench.github.io/

**Comment by clarus** [comment]
Also I think I can makes sense to have several different test environments, as it reduces the biais but maybe this is overkill.

**Comment by maximedenes** [comment]
> Regarding the wider OPAM efforts I'm not sure that is a good idea to involve them in this given the response to my suggestion.

My answer was short because of time constraints. I'm working on separating the website generation and the package index. It is high priority, so I took the simplest approach. We can refine it with a more standardized approach, as you suggested.

I didn't want to dismiss the idea of getting in touch with developers of other tools. You may want to do it from a separate thread, though.

**Comment by ejgallego** [comment]
> I started this project at the end of 2014 and have maintained it since then running it on my server. At that time I do not think I have heard about alternatives on the OCaml side.

Indeed, I guess here we are in a kind of worse-case lack of coordination / publicity scenario; I'm sure the OCaml people didn't know about your tool so now we have two awesome tools which are hard to merge now :(

> Most of the development time was invested into making the platform reliable. I like it like that but it would probably need a notification system, like in Slack as you said @ejgallego .

It would be awesome to post build runs to Coq's gitter, let me know if you need any help for that.

> We could probably use http://check.ocamllabs.io/ for Coq with some adaptations, but I like the stability and independence of https://coq-bench.github.io/

Indeed both are awesome, but after a quick look I have no idea how to try to merge them :/ I'll be happy to help tho, just ping me around here or in Gitter.

I have opened an issue about Coq bench in the main Coq repos to see how we can improve its visibility, c.f.: https://github.com/coq/coq/issues/10418

**Pull request merged**